src/search3_indexer.erl

```diff
@@ -0,0 +1,241 @@
+    #{
+        <<"db_name">> := DbName,
+        <<"ddoc_id">> := DDocId,
+        <<"name">> := IndexName,
+        <<"sig">> := JobSig
+    } = JobData,
+    {ok, Db} = fabric2_db:open(DbName, []),
+    {ok, DDoc} = fabric2_db:open_doc(Db, DDocId),
+    {ok, Index} = search3_util:design_doc_to_index(DDoc, IndexName),
+    Index1 = Index#index{dbname = DbName},
+    HexSig = fabric2_util:to_hex(Index1#index.sig),
+
+    if  HexSig == JobSig -> ok; true ->
+        couch_jobs:finish(undefined, Job, JobData#{
+            error => sig_changed,
+            reason => <<"Design document was modified">>
+        }),
+        exit(normal)
+    end,
+    State = #{
+        txdb => undefined,
+        search_seq => undefined,
+        count => 0,
+        limit => config:get_integer("search3", "change_limit", 100),
+        doc_acc => [],
+        last_seq => undefined,
+        job => Job,
+        job_data => JobData,
+        index => Index1,
+        proc => undefined
+    },
+    update(Db, State).
+
+update(#{} = Db, State) ->
+    try
+        Index = maps:get(index, State),
+        {InitSession, Seq} = search3_rpc:get_update_seq(Index),
+        Proc = get_os_process(Index#index.def_lang),
+        % Start of a new session
+        Index1 = Index#index{session=InitSession},
+        NewState = State#{
+            search_seq => Seq,
+            last_seq => Seq,
+            proc => Proc,
+            index => Index1
+        },
+        try
+            true = proc_prompt(Proc, [<<"add_fun">>, Index1#index.def]),
+            update_int(Db, NewState)
+        after
+            ret_os_process(Proc)
+        end
+    catch error:database_does_not_exist ->
+        #{db_prefix := DbPrefix} = Db,
+        couch_log:error("Db - ~p database does not exist", [DbPrefix]),
+        erlang:error(database_does_not_exist)
+    end.
+
+update_int(#{} = Db, State) ->
+    State4 = fabric2_fdb:transactional(Db, fun(TxDb) ->
+        State1 = maps:put(tx_db, TxDb, State),
+        {ok, State2} = fold_changes(State1),
+        #{
+            count := Count,
+            limit := Limit,
+            doc_acc := DocAcc,
+            last_seq := LastSeq,
+            index := Index,
+            proc := Proc
+        } = State2,
+
+        % 1) purge_seq is empty for now, subsequent releases will support this
+        % 2) We are indexing inside a transaction, which has a timeout of 5s.
+        % This can potentially be a problem for large documents. We should
+        % revisit this design later.
+        Session = try
+             index_docs(Index, Proc, LastSeq, <<>>, DocAcc)
+        catch error:session_mismatch ->
+            % if there is a session mismatch, we just finish this job and
+            % restart it again
+            report_progress(State2, finished),
+            exit(normal)
+        end,
+
+        % Update the session after each update
+        Index1 = Index#index{session = Session},
```

Comment:
explain? This would appear to break the whole point of session; that it must not change during the get_update_seq/updates/set_update_seq workflow.

Reply:
or your comment is just really bad?